src/session/mpi.rs

```diff
@@ -125,28 +139,40 @@ impl SessionMPI {
 
         for provider in &self.provider_sessions {
             provider
-                .download(blob_id, "app".to_owned(), ResourceFormat::Tar)
+                .download(blob_id, app_path.clone(), ResourceFormat::Tar)
                 .context("downloading file")?;
 
+            // If we create a ProviderSession per provider, every session
+            // gets a unique identifier. This means that the resulting executable
+            // resides in a different directory on each of the provider nodes,
+            // which causes mpirun to fail.
+            // As a workaround, we provide a symlink to the /tmp directory
+            // in the image and put the resulting binary there.
+
+            // For the CMake backend we use the EXECUTABLE_OUTPUT_PATH CMake variable
+            // For the Make backend we just move the file around
+
             let cmake_cmd = Command::Exec {
                 executable: "cmake/bin/cmake".to_owned(),
                 args: vec![
-                    "app",
-                    "-DCMAKE_C_COMPILER=mpicc",
-                    "-DCMAKE_CXX_COMPILER=mpicxx",
-                    "-DCMAKE_BUILD_TYPE=Release",
-                ]
-                .into_iter()
-                .map(String::from)
-                .collect(),
+                    app_path.clone(),
+                    "-DCMAKE_C_COMPILER=mpicc".to_owned(),
+                    "-DCMAKE_CXX_COMPILER=mpicxx".to_owned(),
+                    "-DCMAKE_BUILD_TYPE=Release".to_owned(),
+                    "-DEXECUTABLE_OUTPUT_PATH=tmp".to_owned(), // TODO fix path for Make
+                ],
+            };
+            let mv_cmd = Command::Exec {
+                executable: "mv".to_owned(),
+                args: vec!["app/".to_owned() + progname, "tmp/".to_owned()],
```

Comment:
Consider using Path.join and reusing `app_path`

Reply:
Yeah, the thing is, neither `Path` nor `PathBuf` coerces to `str` or `String` automatically. You either convert them to `Cow<str>` which then coerces to UTF8-only string (yeah, alas it's a lossy operation, where each nonUTF8 is substituted with garbage), or you convert to `Option<&str>` which forces you to handle errors in some way.

So yeah, good idea about using `Path.join` to automatically intersperse with "/" but it might reduce the clarity. There is an alternative however, in the `itertools` crate called (surprise, surprise) [join](https://docs.rs/itertools/0.8.0/itertools/fn.join.html):

```rust
assert_eq!(join(&["app", progname], "/"), "app/<progname>");
```

Reply:
We don't have to depend on 3rd party libraries - there's [join](https://doc.rust-lang.org/std/slice/trait.SliceConcatExt.html#tymethod.join) in stdlib.

Thanks for noticing that.

Reply:
Thanks for noticing that.

We don't have to depend on 3rd party libraries - there's [join](https://doc.rust-lang.org/std/slice/trait.SliceConcatExt.html#tymethod.join) in stdlib. Take look if it's ok now.